File: src/cathedral/store.py
from __future__ import annotations

import os
import re
import shutil
import subprocess
import sys
from dataclasses import asdict, dataclass
from datetime import datetime
from importlib.resources import files
from pathlib import Path
from typing import BinaryIO, Iterable

import tomllib


STORE_MARKERS = ("Index.md", "nodes", "inbox", "archive", "meta/Guidelines.md", "meta/Sources.md")
ITEM_RE = re.compile(r"^\d{4}-\d{2}-\d{2}-[a-z0-9]+(?:-[a-z0-9]+)*(?:-\d{4})?(?:-\d+)?$")


class CathedralError(Exception):
    """An expected error suitable for display without a traceback."""


@dataclass(frozen=True)
class IngestedItem:
    name: str
    path: str
    kind: str


def is_store(path: Path) -> bool:
    return all((path / marker).exists() for marker in STORE_MARKERS)


def find_store(explicit: str | os.PathLike[str] | None = None) -> Path:
    if explicit:
        candidate = Path(explicit).expanduser().resolve()
        if not is_store(candidate):
            raise CathedralError(f"not a Cathedral store: {candidate}")
        return candidate

    configured = os.environ.get("CATHEDRAL_STORE")
    if configured:
        candidate = Path(configured).expanduser().resolve()
        if not is_store(candidate):
            raise CathedralError(f"CATHEDRAL_STORE is not a Cathedral store: {candidate}")
        return candidate

    for candidate in (Path.cwd(), *Path.cwd().parents):
        if is_store(candidate):
            return candidate.resolve()
    raise CathedralError("no Cathedral store found; use --store PATH or run inside a store")


def template_text(name: str) -> str:
    return files("cathedral").joinpath("templates", name).read_text(encoding="utf-8")


def init_store(path: Path, operator: str, codex_command: str, initialize_git: bool = True) -> dict[str, object]:
    path = path.expanduser().resolve()
    if path.exists() and any(path.iterdir()):
        raise CathedralError(f"directory is not empty: {path}")
    path.mkdir(parents=True, exist_ok=True)
    for directory in ("nodes", "inbox", "archive", "meta"):
        (path / directory).mkdir()

    (path / "Index.md").write_text(
        "# Index\n\nA dense, current map of the memories in this store.\n",
        encoding="utf-8",
    )
    (path / "meta" / "Guidelines.md").write_text(template_text("Guidelines.md"), encoding="utf-8")
    (path / "meta" / "Consolidation.md").write_text(template_text("Consolidation.md"), encoding="utf-8")
    (path / "meta" / "Sources.md").write_text(
        "# Sources\n\n"
        f"## {operator}\n"
        "- role: operator\n"
        "- salience: highest — their statements, decisions, and syntheses are remembered preferentially in any context.\n",
        encoding="utf-8",
    )
    config = template_text("Config.toml").replace("codex_command = \"codex\"", f"codex_command = {toml_string(codex_command)}")
    (path / "meta" / "Config.toml").write_text(config, encoding="utf-8")

    git_initialized = False
    if initialize_git and not inside_git_repository(path):
        result = subprocess.run(
            ["git", "init", "--quiet", str(path)],
            text=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
        if result.returncode != 0:
            raise CathedralError(f"could not initialize Git repository: {result.stderr.strip()}")
        git_initialized = True
    return {"store": str(path), "operator": operator, "codex_command": codex_command, "git_initialized": git_initialized}


def toml_string(value: str) -> str:
    return '"' + value.replace("\\", "\\\\").replace('"', '\\"') + '"'


def inside_git_repository(path: Path) -> bool:
    probe = path if path.exists() else path.parent
    result = subprocess.run(
        ["git", "-C", str(probe), "rev-parse", "--is-inside-work-tree"],
        stdout=subprocess.DEVNULL,
        stderr=subprocess.DEVNULL,
    )
    return result.returncode == 0


def load_config(store: Path) -> dict[str, object]:
    config_path = store / "meta" / "Config.toml"
    if not config_path.exists():
        return {"codex_command": "codex", "max_recall_nodes": 6}
    try:
        with config_path.open("rb") as stream:
            config = tomllib.load(stream)
    except (OSError, tomllib.TOMLDecodeError) as error:
        raise CathedralError(f"invalid {config_path}: {error}") from error
    return {"codex_command": "codex", "max_recall_nodes": 6, **config}


def slugify(value: str) -> str:
    value = value.lower().strip()
    value = re.sub(r"[^a-z0-9]+", "-", value).strip("-")
    if not value:
        raise CathedralError("slug must contain a letter or number")
    return value


def available_item_path(inbox: Path, slug: str, now: datetime | None = None) -> Path:
    now = now or datetime.now().astimezone()
    base = f"{now:%Y-%m-%d}-{slugify(slug)}"
    candidate = inbox / base
    if not candidate.exists():
        return candidate
    candidate = inbox / f"{base}-{now:%H%M}"
    if not candidate.exists():
        return candidate
    suffix = 2
    while (inbox / f"{base}-{now:%H%M}-{suffix}").exists():
        suffix += 1
    return inbox / f"{base}-{now:%H%M}-{suffix}"


def ingest(
    store: Path,
    sources: Iterable[str],
    slug: str | None = None,
    stdin: BinaryIO | None = None,
) -> list[IngestedItem]:
    source_list = list(sources)
    if slug and len(source_list) != 1:
        raise CathedralError("--slug requires exactly one input")
    if source_list.count("-") > 1:
        raise CathedralError("stdin may only be ingested once per command")

    results: list[IngestedItem] = []
    for source_name in source_list:
        if source_name == "-":
            item_path = available_item_path(store / "inbox", slug or "stdin")
            input_stream = stdin or sys.stdin.buffer
            item_path.write_bytes(input_stream.read())
            kind = "file"
        else:
            source = Path(source_name).expanduser()
            if not source.exists():
                raise CathedralError(f"input does not exist: {source}")
            item_slug = slug or source.stem or source.name
            item_path = available_item_path(store / "inbox", item_slug)
            if source.is_dir():
                shutil.copytree(source, item_path, copy_function=shutil.copy2)
                kind = "directory"
            elif source.is_file():
                shutil.copy2(source, item_path)
                kind = "file"
            else:
                raise CathedralError(f"input is not a regular file or directory: {source}")
        results.append(IngestedItem(item_path.name, str(item_path), kind))
    return results


def list_items(directory: Path) -> list[dict[str, object]]:
    result = []
    for path in sorted(directory.iterdir()):
        stat = path.stat()
        result.append(
            {
                "name": path.name,
                "path": str(path),
                "kind": "directory" if path.is_dir() else "file",
                "size": tree_size(path),
                "modified": int(stat.st_mtime),
            }
        )
    return result


def tree_size(path: Path) -> int:
    if path.is_file():
        return path.stat().st_size
    return sum(child.stat().st_size for child in path.rglob("*") if child.is_file())


def read_item(path: Path) -> str:
    if not path.exists():
        raise CathedralError(f"item does not exist: {path.name}")
    paths = [path] if path.is_file() else sorted(child for child in path.rglob("*") if child.is_file())
    sections: list[str] = []
    for child in paths:
        try:
            content = child.read_text(encoding="utf-8")
        except UnicodeDecodeError:
            content = f"[binary file: {child.stat().st_size} bytes]"
        label = child.name if path.is_file() else str(child.relative_to(path))
        sections.append(f"## {label}\n\n{content.rstrip()}\n")
    return "\n".join(sections)


def serialized_items(items: Iterable[IngestedItem]) -> list[dict[str, object]]:
    return [asdict(item) for item in items]

File: store.go
package main

import (
	"embed"
	"encoding/json"
	"errors"
	"fmt"
	"io"
	"io/fs"
	"os"
	"os/exec"
	"path/filepath"
	"regexp"
	"sort"
	"strconv"
	"strings"
	"time"
	"unicode/utf8"
)

//go:embed research/doc_guidelines.md research/prompt_consolidation.md
var researchFiles embed.FS

var itemPattern = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}-[a-z0-9]+(?:-[a-z0-9]+)*(?:-\d{4})?(?:-\d+)?$`)

type ingestedItem struct {
	Name string `json:"name"`
	Path string `json:"path"`
	Kind string `json:"kind"`
}

type listedItem struct {
	Name     string `json:"name"`
	Path     string `json:"path"`
	Kind     string `json:"kind"`
	Size     int64  `json:"size"`
	Modified int64  `json:"modified"`
}

type config struct {
	CodexCommand         string
	CodexModel           string
	CodexReasoningEffort string
	MaxRecallNodes       int
}

const (
	defaultCodexCommand         = "codex"
	defaultCodexModel           = "gpt-5.6-terra"
	defaultCodexReasoningEffort = "high"
)

func isStore(path string) bool {
	markers := []string{"Index.md", "nodes", "inbox", "archive", "meta/Guidelines.md", "meta/Sources.md"}
	for _, marker := range markers {
		if _, err := os.Stat(filepath.Join(path, marker)); err != nil {
			return false
		}
	}
	return true
}

func findStore(explicit string) (string, error) {
	if explicit != "" {
		path, err := absolutePath(explicit)
		if err != nil {
			return "", err
		}
		if !isStore(path) {
			return "", fmt.Errorf("not a Cathedral store: %s", path)
		}
		return path, nil
	}
	if configured := os.Getenv("CATHEDRAL_STORE"); configured != "" {
		path, err := absolutePath(configured)
		if err != nil {
			return "", err
		}
		if !isStore(path) {
			return "", fmt.Errorf("CATHEDRAL_STORE is not a Cathedral store: %s", path)
		}
		return path, nil
	}
	current, err := os.Getwd()
	if err != nil {
		return "", err
	}
	for {
		if isStore(current) {
			return current, nil
		}
		parent := filepath.Dir(current)
		if parent == current {
			break
		}
		current = parent
	}
	return "", errors.New("no Cathedral store found; use --store PATH or run inside a store")
}

func absolutePath(path string) (string, error) {
	if strings.HasPrefix(path, "~/") {
		home, err := os.UserHomeDir()
		if err != nil {
			return "", err
		}
		path = filepath.Join(home, path[2:])
	}
	return filepath.Abs(path)
}

func initializeStore(path, operator, codexCommand string, initializeGit bool) (map[string]any, error) {
	return initializeStoreWithConfig(path, operator, codexCommand, defaultCodexModel, defaultCodexReasoningEffort, initializeGit)
}

func initializeStoreWithConfig(path, operator, codexCommand, codexModel, codexReasoningEffort string, initializeGit bool) (map[string]any, error) {
	path, err := absolutePath(path)
	if err != nil {
		return nil, err
	}
	if entries, readErr := os.ReadDir(path); readErr == nil && len(entries) > 0 {
		return nil, fmt.Errorf("directory is not empty: %s", path)
	} else if readErr != nil && !os.IsNotExist(readErr) {
		return nil, readErr
	}
	for _, directory := range []string{"nodes", "inbox", "archive", "meta"} {
		if err := os.MkdirAll(filepath.Join(path, directory), 0o755); err != nil {
			return nil, err
		}
	}
	guidelines, err := researchFiles.ReadFile("research/doc_guidelines.md")
	if err != nil {
		return nil, err
	}
	consolidation, err := researchFiles.ReadFile("research/prompt_consolidation.md")
	if err != nil {
		return nil, err
	}
	files := map[string][]byte{
		"Index.md":              []byte("# Index\n\nA dense, current map of the memories in this store.\n"),
		"meta/Guidelines.md":    guidelines,
		"meta/Consolidation.md": consolidation,
		"meta/Sources.md":       []byte(fmt.Sprintf("# Sources\n\n## %s\n- role: operator\n- salience: highest — their statements, decisions, and syntheses are remembered preferentially in any context.\n", operator)),
		"meta/Config.toml":      []byte(fmt.Sprintf("# Command prefix used to invoke Codex. Cathedral appends `exec` and its required\n# non-interactive flags. Shell quoting is supported, but shell syntax is not.\ncodex_command = %s\n\n# Model and reasoning effort for Codex consolidation. They can be overridden for\n# one run with `cathedral consolidate --model MODEL --reasoning-effort EFFORT`.\ncodex_model = %s\ncodex_reasoning_effort = %s\n\n# Default upper bound for deterministic recall bundles.\nmax_recall_nodes = 6\n", strconv.Quote(codexCommand), strconv.Quote(codexModel), strconv.Quote(codexReasoningEffort))),
	}
	for name, contents := range files {
		if err := os.WriteFile(filepath.Join(path, name), contents, 0o644); err != nil {
			return nil, err
		}
	}
	gitInitialized := false
	if initializeGit && !insideGitRepository(path) {
		command := exec.Command("git", "init", "--quiet", path)
		if output, err := command.CombinedOutput(); err != nil {
			return nil, fmt.Errorf("could not initialize Git repository: %s", strings.TrimSpace(string(output)))
		}
		gitInitialized = true
		if err := commitInitialStore(path); err != nil {
			return nil, err
		}
	}
	return map[string]any{"store": path, "operator": operator, "codex_command": codexCommand, "codex_model": codexModel, "codex_reasoning_effort": codexReasoningEffort, "git_initialized": gitInitialized, "initial_commit": gitInitialized}, nil
}

func commitInitialStore(path string) error {
	commands := [][]string{
		{"add", "-A"},
		{"-c", "user.email=cathedral@localhost", "-c", "user.name=Cathedral", "commit", "--quiet", "-m", "Initialize Cathedral store"},
	}
	for _, arguments := range commands {
		command := exec.Command("git", append([]string{"-C", path}, arguments...)...)
		if output, err := command.CombinedOutput(); err != nil {
			return fmt.Errorf("could not create initial Cathedral commit: %s", strings.TrimSpace(string(output)))
		}
	}
	return nil
}

func insideGitRepository(path string) bool {
	command := exec.Command("git", "-C", path, "rev-parse", "--is-inside-work-tree")
	command.Stdout = io.Discard
	command.Stderr = io.Discard
	return command.Run() == nil
}

func loadConfig(store string) (config, error) {
	result := config{CodexCommand: defaultCodexCommand, CodexModel: defaultCodexModel, CodexReasoningEffort: defaultCodexReasoningEffort, MaxRecallNodes: 6}
	contents, err := os.ReadFile(filepath.Join(store, "meta", "Config.toml"))
	if os.IsNotExist(err) {
		return result, nil
	}
	if err != nil {
		return result, err
	}
	commandPattern := regexp.MustCompile(`(?m)^\s*codex_command\s*=\s*("(?:\\.|[^"])*")\s*$`)
	modelPattern := regexp.MustCompile(`(?m)^\s*codex_model\s*=\s*("(?:\\.|[^"])*")\s*$`)
	reasoningPattern := regexp.MustCompile(`(?m)^\s*codex_reasoning_effort\s*=\s*("(?:\\.|[^"])*")\s*$`)
	maxPattern := regexp.MustCompile(`(?m)^\s*max_recall_nodes\s*=\s*(\d+)\s*$`)
	if match := commandPattern.FindSubmatch(contents); match != nil {
		if value, parseErr := strconv.Unquote(string(match[1])); parseErr == nil {
			result.CodexCommand = value
		} else {
			return result, fmt.Errorf("invalid meta/Config.toml codex_command: %w", parseErr)
		}
	}
	if match := modelPattern.FindSubmatch(contents); match != nil {
		value, parseErr := strconv.Unquote(string(match[1]))
		if parseErr != nil {
			return result, fmt.Errorf("invalid meta/Config.toml codex_model: %w", parseErr)
		}
		if strings.TrimSpace(value) == "" {
			return result, errors.New("invalid meta/Config.toml codex_model: must not be empty")
		}
		result.CodexModel = value
	}
	if match := reasoningPattern.FindSubmatch(contents); match != nil {
		value, parseErr := strconv.Unquote(string(match[1]))
		if parseErr != nil {
			return result, fmt.Errorf("invalid meta/Config.toml codex_reasoning_effort: %w", parseErr)
		}
		if strings.TrimSpace(value) == "" {
			return result, errors.New("invalid meta/Config.toml codex_reasoning_effort: must not be empty")
		}
		result.CodexReasoningEffort = value
	}
	if match := maxPattern.FindSubmatch(contents); match != nil {
		value, _ := strconv.Atoi(string(match[1]))
		if value < 1 {
			return result, errors.New("invalid meta/Config.toml max_recall_nodes: must be positive")
		}
		result.MaxRecallNodes = value
	}
	return result, nil
}

func slugify(value string) (string, error) {
	value = strings.ToLower(strings.TrimSpace(value))
	value = regexp.MustCompile(`[^a-z0-9]+`).ReplaceAllString(value, "-")
	value = strings.Trim(value, "-")
	if value == "" {
		return "", errors.New("slug must contain a letter or number")
	}
	return value, nil
}

func availableItemPath(inbox, slug string, now time.Time) (string, error) {
	slug, err := slugify(slug)
	if err != nil {
		return "", err
	}
	base := now.Format("2006-01-02") + "-" + slug
	candidate := filepath.Join(inbox, base)
	if !pathExists(candidate) {
		return candidate, nil
	}
	candidate = filepath.Join(inbox, base+"-"+now.Format("1504"))
	if !pathExists(candidate) {
		return candidate, nil
	}
	for suffix := 2; ; suffix++ {
		candidate = filepath.Join(inbox, fmt.Sprintf("%s-%s-%d", base, now.Format("1504"), suffix))
		if !pathExists(candidate) {
			return candidate, nil
		}
	}
}

func ingestInputs(store string, sources []string, slug string, stdin io.Reader, now time.Time) ([]ingestedItem, error) {
	if slug != "" && len(sources) != 1 {
		return nil, errors.New("--slug requires exactly one input")
	}
	result := make([]ingestedItem, 0, len(sources))
	for _, sourceName := range sources {
		kind := "file"
		itemSlug := slug
		if sourceName == "-" {
			if itemSlug == "" {
				itemSlug = "stdin"
			}
			path, err := availableItemPath(filepath.Join(store, "inbox"), itemSlug, now)
			if err != nil {
				return nil, err
			}
			file, err := os.OpenFile(path, os.O_CREATE|os.O_EXCL|os.O_WRONLY, 0o644)
			if err != nil {
				return nil, err
			}
			_, copyErr := io.Copy(file, stdin)
			closeErr := file.Close()
			if copyErr != nil {
				return nil, copyErr
			}
			if closeErr != nil {
				return nil, closeErr
			}
			result = append(result, ingestedItem{filepath.Base(path), path, kind})
			continue
		}
		source, err := absolutePath(sourceName)
		if err != nil {
			return nil, err
		}
		info, err := os.Stat(source)
		if err != nil {
			return nil, fmt.Errorf("input does not exist: %s", sourceName)
		}
		if itemSlug == "" {
			itemSlug = strings.TrimSuffix(filepath.Base(source), filepath.Ext(source))
		}
		destination, err := availableItemPath(filepath.Join(store, "inbox"), itemSlug, now)
		if err != nil {
			return nil, err
		}
		if info.IsDir() {
			kind = "directory"
			err = copyDirectory(source, destination)
		} else if info.Mode().IsRegular() {
			err = copyFile(source, destination, info)
		} else {
			return nil, fmt.Errorf("input is not a regular file or directory: %s", sourceName)
		}
		if err != nil {
			return nil, err
		}
		result = append(result, ingestedItem{filepath.Base(destination), destination, kind})
	}
	return result, nil
}

func copyDirectory(source, destination string) error {
	return filepath.WalkDir(source, func(path string, entry fs.DirEntry, walkErr error) error {
		if walkErr != nil {
			return walkErr
		}
		relative, err := filepath.Rel(source, path)
		if err != nil {
			return err
		}
		target := filepath.Join(destination, relative)
		info, err := entry.Info()
		if err != nil {
			return err
		}
		if entry.IsDir() {
			return os.MkdirAll(target, info.Mode().Perm())
		}
		if !entry.Type().IsRegular() {
			return fmt.Errorf("input contains unsupported file type: %s", path)
		}
		return copyFile(path, target, info)
	})
}

func copyFile(source, destination string, info fs.FileInfo) error {
	input, err := os.Open(source)
	if err != nil {
		return err
	}
	defer input.Close()
	output, err := os.OpenFile(destination, os.O_CREATE|os.O_EXCL|os.O_WRONLY, info.Mode().Perm())
	if err != nil {
		return err
	}
	if _, err = io.Copy(output, input); err != nil {
		output.Close()
		return err
	}
	if err = output.Close(); err != nil {
		return err
	}
	return os.Chtimes(destination, info.ModTime(), info.ModTime())
}

func listItems(directory string) ([]listedItem, error) {
	entries, err := os.ReadDir(directory)
	if err != nil {
		return nil, err
	}
	result := make([]listedItem, 0, len(entries))
	for _, entry := range entries {
		path := filepath.Join(directory, entry.Name())
		info, err := entry.Info()
		if err != nil {
			return nil, err
		}
		size, err := treeSize(path)
		if err != nil {
			return nil, err
		}
		kind := "file"
		if entry.IsDir() {
			kind = "directory"
		}
		result = append(result, listedItem{entry.Name(), path, kind, size, info.ModTime().Unix()})
	}
	sort.Slice(result, func(i, j int) bool { return result[i].Name < result[j].Name })
	return result, nil
}

func treeSize(path string) (int64, error) {
	var size int64
	err := filepath.WalkDir(path, func(child string, entry fs.DirEntry, err error) error {
		if err != nil {
			return err
		}
		if entry.Type().IsRegular() {
			info, err := entry.Info()
			if err != nil {
				return err
			}
			size += info.Size()
		}
		return nil
	})
	return size, err
}

func readItem(path string) (string, error) {
	info, err := os.Stat(path)
	if err != nil {
		return "", fmt.Errorf("item does not exist: %s", filepath.Base(path))
	}
	paths := []string{path}
	if info.IsDir() {
		paths = nil
		err = filepath.WalkDir(path, func(child string, entry fs.DirEntry, err error) error {
			if err == nil && entry.Type().IsRegular() {
				paths = append(paths, child)
			}
			return err
		})
		if err != nil {
			return "", err
		}
		sort.Strings(paths)
	}
	var builder strings.Builder
	for index, child := range paths {
		contents, err := os.ReadFile(child)
		if err != nil {
			return "", err
		}
		label := filepath.Base(child)
		if info.IsDir() {
			label, _ = filepath.Rel(path, child)
		}
		if index > 0 {
			builder.WriteByte('\n')
		}
		fmt.Fprintf(&builder, "## %s\n\n", label)
		if json.Valid(contents) || isText(contents) {
			builder.Write(contents)
			if len(contents) == 0 || contents[len(contents)-1] != '\n' {
				builder.WriteByte('\n')
			}
		} else {
			fmt.Fprintf(&builder, "[binary file: %d bytes]\n", len(contents))
		}
	}
	return builder.String(), nil
}

func isText(contents []byte) bool {
	if !utf8.Valid(contents) {
		return false
	}
	for _, value := range contents {
		if value == 0 {
			return false
		}
	}
	return true
}

func pathExists(path string) bool {
	_, err := os.Stat(path)
	return err == nil
}
